File: 19_数学/采样/拒绝采样/519. 随机翻转矩阵.ts
import { randint } from '../randint'

// 尽量最少调用随机函数 Math.random()，并且优化时间和空间复杂度。
class Solution {
  private total: number
  private rows: number
  private cols: number
  private black: Map<number, number>

  constructor(m: number, n: number) {
    this.total = m * n
    this.rows = m
    this.cols = n
    this.black = new Map<number, number>()
  }

  /**
   * 均匀随机的将矩阵中的 0 变为 1，
   * 并返回该值的位置下标 [row_id,col_id]
   * @description
   * 一次调用random方法
   * 每次flip后都将flip的值记录到黑名单中，并将其映射到最后一个数上
   */
  flip(): number[] {
    if (this.total <= 0) return []
    this.total--
    const choose = randint(0, this.total)
    const white = this.getWhite(choose)
    // choose 映射到另一个白名单的数
    this.black.set(choose, this.getWhite(this.total))
    return [~~(white / this.cols), white % this.cols]
  }

  /**
   * 将所有的值都重新置为 0
   */
  reset(): void {
    this.total = this.rows * this.cols
    this.black.clear()
  }

  private getWhite(num: number): number {
    return this.black.get(num) || num
  }
}

const s = new Solution(2, 3)
console.log(s.flip())
s.reset()
console.log(s.flip())
export {}

File: 19_数学/采样/拒绝采样/519. 随机翻转矩阵.py
from random import randint
from typing import List


class Solution:
    """
    随机选取一个满足 matrix[i][j] == 0 的下标 (i, j) ，并将它的值变为 1 。
    所有满足 matrix[i][j] == 0 的下标 (i, j) 被选取的概率应当均等。
    尽量最少调用内置的随机函数，并且优化时间和空间复杂度。
    """

    def __init__(self, m: int, n: int):
        self.black = dict()
        self.row, self.col = m, n
        self.remain = m * n

    def flip(self) -> List[int]:
        """返回一个满足 matrix[i][j] == 0 的随机下标 [i, j] ，并将其对应格子中的值变为 1

        每次flip后 都将flip的值记录到黑名单 并映射到另一个不是黑名单的下标(这里指最后一个数)
        """
        if self.remain == 0:
            return []
        self.remain -= 1
        rand = randint(0, self.remain)
        res = self.black.get(rand, rand)  # rand在或不在都可以选
        self.black[rand] = self.black.get(self.remain, self.remain)
        return [res // self.col, res % self.col]

    def reset(self) -> None:
        """将矩阵中所有的值重置为 0"""
        self.remain = self.row * self.col
        self.black.clear()


# Your Solution object will be instantiated and called as such:
# obj = Solution(m, n)
# param_1 = obj.flip()
# obj.reset()
